### PR TITLE
AMDGPU: Update pattern matching from "x&(-1>>(32-y))" to "bfe x, 0, y"

### PR DESCRIPTION
**Pull request by changpeng**
It is not correct to lower "x&(-1>>(32-y))" to "bfe x, 0, y". When y equals 32, "-1" is not shifted, so x&(-1>>(32-32) is still x, but "bfe x, 0, 32" is 0. However, if we know y is at most of 5 bits (< 32), we can still do the pattern matching.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/extract-lowbits.ll [hunk 1]
```diff
@@ -99,11 +99,21 @@ define i32 @bzhi32_b4_commutative(i32 %val, i32 %numlowbits) nounwind {
 ; ---------------------------------------------------------------------------- ;
 
 define i32 @bzhi32_c0(i32 %val, i32 %numlowbits) nounwind {
-; GCN-LABEL: bzhi32_c0:
-; GCN:       ; %bb.0:
-; GCN-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; GCN-NEXT:    v_bfe_u32 v0, v0, 0, v1
-; GCN-NEXT:    s_setpc_b64 s[30:31]
+; SI-LABEL: bzhi32_c0:
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:102]: This should be OK if numlowbits is known < bitwidth. This case is canonical, so you should bring back the predicate from the last pr

Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:102]: Frankly speaking, I have difficulty in understanding what is canonical here.  Is the following IR still canonical?
  And do you mean we can check MaxActiveBits to allow the pattern?  Thanks.
    
  %low5bits      = and i32 %numlowbits, 31
  %numhighbits = sub i32 32, %numlowbits
  %mask = lshr i32 -1, %numhighbits
  %masked = and i32 %mask, %val

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:102]: run it through instcombine, and computeKnownBits as you did before

Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:102]: > Run it through instcombine. Whatever it produces is canonical. and computeKnownBits as you did before
Running instcombine like this?
./opt -passes=instcombine canon.ll -S

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:102]: yes

**Comment by jayfoad** [comment]
Mismatched parens in both summary and description: x&(-1>>(32-y)

**Comment by changpeng** [comment]
> Mismatched parens in both summary and description: x&(-1>>(32-y)
Thanks!

**Comment by rovka** [comment]
Hi, I like your patch but I have a couple of questions:
* Where does it say that `v_bfe_u32 x, 0, 32` is 0? The description in [the RDNA3 ISA](https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf) says `D0.u = ((S0.u >> S1.u[4 : 0].u) & 32'U((1 << S2.u[4 : 0].u) - 1))`, which is unfortunately not very specific about edge cases. I would interpret it as large values cause the 1 to be shifted out, then when we subtract 1 we get 0xFFF...F, so when we do & with it we essentially get x. If the hardware actually returns 0, we should probably talk to folks to make that clear in the ISA.
* While you're at it, what do `v_bfe_i32` and `s_bfe_*` do and do we have proper tests for them?

**Comment by changpeng** [comment]
[AMD Official Use Only - AMD Internal Distribution Only]

Right. So bfe x, 0, 32 is the same as bfe x, 0, 0 and the result is 0.  But x & (-1 >> 0) equals x.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Review by arsenm — APPROVED** [review]

**Pull request merged**